src/Models/HydrostaticFreeSurfaceModels/implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,77 @@
+using Oceananigans.Solvers
+using Oceananigans.Operators
+
+struct ImplicitFreeSurfaceSolver{S}
+    solver   :: S
+end
+
+function ImplicitFreeSurfaceSolver(arch, template_field, 
+                                   vertically_integrated_lateral_face_areas;
+                                   Amatrix_operator=nothing, 
+                                   maxit=nothing, 
+                                   tol=nothing)
+       
+    ## This should probably end up in some operator and grid generic auxilliaries
+    ## NEED TO SORT OUT HALO FOR PERIODIC CASE
+    Ax_baro = vertically_integrated_lateral_face_areas.Ax
+    Ay_baro = vertically_integrated_lateral_face_areas.Ay
+    @inline Ax_∂xᶠᵃᵃ_baro(i, j, k, grid, c) = Ax_baro[i, j, 1] * ∂xᶠᵃᵃ(i, j, 1, grid, c)
+    @inline Ay_∂yᵃᶠᵃ_baro(i, j, k, grid, c) = Ay_baro[i, j, 1] * ∂yᵃᶠᵃ(i, j, 1, grid, c)
+    @inline function ∇²_baro(i, j, k, grid, c)
+       return  δxᶜᵃᵃ(i, j, 1, grid, Ax_∂xᶠᵃᵃ_baro, c) +
+               δyᵃᶜᵃ(i, j, 1, grid, Ay_∂yᵃᶠᵃ_baro, c)
+    end
+
+    @kernel function implicit_η!(grid, f, implicit_η_f)
+        i, j = @index(Global, NTuple)
+        ### Not sure what to call this
+        ### it is for left hand side operator in
+        ### ( ∇ʰ⋅H∇ʰ - 1/gΔt² ) ηⁿ⁺¹ = 1/(gΔt) ∇ʰH U̅ˢᵗᵃʳ - 1/(gΔt²) ηⁿ
+        ### written in a discrete finite volume form in which the equation
+        ### is arranged to ensure a symmtric form
+        ### e.g.
+        ### 
+        ### δⁱÂʷ∂ˣηⁿ⁺¹ + δʲÂˢ∂ʸηⁿ⁺¹ - 1/gΔt² Aᶻηⁿ⁺¹ =
```

Comment:
This is what we end up solving.

Reply:
Is it suspicious that there's a `Δt²` here, and then a seemingly spurious factor of `Δt` added elsewhere? In other words, it seems we are scaling by `η` by `Δt`?

Reply:
That is also in 2.15 of this https://mitgcm.readthedocs.io/en/latest/algorithm/algorithm.html#pressure-method-with-implicit-linear-free-surface . I got the same thing when I went through it to check.

Reply:
Hmm, I think I'm going to retract my comment!

Reply:
It's the same here:

https://github.com/CliMA/Oceananigans.jl/blob/a211d9d0f0ddf3f66ca9fe4319f47aac1e9e6164/src/Models/IncompressibleModels/pressure_correction.jl#L30-L36

and the reason is that this is really a pseudo time-step. Ok, I'm ok!__

Reply:
OK - that explains why I had to add a Δt in https://github.com/CliMA/Oceananigans.jl/blob/8586d47ee7ac5d4351334848b7cccfce67e88384/src/Models/HydrostaticFreeSurfaceModels/barotropic_pressure_correction.jl#L33

Still be good to write out the discrete variable grid form in the docs.